Implementing tracking of dependency information for test methods

selenareneephillips opened the pull request:
Changes:

Updates to execution state tracker, execution state listener and sample classes in parallelization test suite to track dependency information for test methods.

juherr commented on src/test/java/test/thread/parallelization/TestNgRunStateTracker.java:
```diff
@@ -428,6 +429,60 @@ public static Long getTestListenerFinishThreadId(String suiteName, String testNa
+
+    public static Multimap<Object, EventLog> getTestMethodEventLogsForMethodsDependedOn(String suiteName, String
+            testName, String className, String methodName) {
+        //System.out.println("\nProcessing " + className + ", " + methodName);
```
Avoid commented code if possible

selenareneephillips commented:
I added another commit. I have commented out the logger statements and replaced them with System.out.println statements so that failures in the parallelization test suite on Travis CI can be debugged.

selenareneephillips commented:
I have figured out why the parallelization tests are failing on Travis CI. I designed the verification for true method parallelization around the expectation that the order of the event log would indicate when events occurred.  The expectation is that a block of methods executing in parallel would result in an event log in which their start events would all be logged, followed by all their event execution events and finally their method pass events. Following that I expected a similar block of event logs for the next batch of methods to execute in parallel.

On my local system, this is how it _always_ works. On Travis CI, I see things like the example below happening. In this example, all the start events are logged first, however, some of the test method pass events for some methods are logged before the test method execution events of other methods. There is a larger variation between the time stamps of test method start events on the Travis CI system than there is on my local system. I think it is the cumulative effect of this larger variation that results in some of the test method pass events getting logged before all the test method execution events are logged for this block of methods.

Also interesting is that this test involves a factory which produced two instances of the test class. It seems that there is a small delay of milliseconds between the method of one instance and the same method on the other instance. 

testMethodA on instance 1 starts a few milliseconds before testMethodA on instance 2
testMethodB on instance 1 starts a few milliseconds before testMethodB on instance 2 

.... and so on.

I also noticed that both instances of testMethodA start a few milliseconds before the instances of testMethodB, and both instances of testMethodB start a few milliseconds before the instances of testMethodC, and so on.

So, my question is -- this has nothing to do with TestNG, right? All of this is due to Java's handling of the threads and it's not anything TestNG has control over? If so, then I can refactor my code pretty easily to process the logs differently. 

**TEST CASE 5, Scenario 1**

This test scenario consists of a single suite with a single test consisting of a factory that provides two instances of a single test class with five methods. There are no dependencies or data providers.

**Event logs extracted from event log list between index 0 and index 9 should be the test method execution event logs for a block of 10 simultaneously executing methods**

Event logged at index 0: EventLog{Event: LISTENER_TEST_METHOD_START, Class: TestClassAFiveMethodsWithNoDepsSample, Class instance hash code: 473286673, Method name: testMethodA, Time of event: 2017-09-10 16:55:41.617, Thread ID: 5446}
Event logged at index 1: EventLog{Event: LISTENER_TEST_METHOD_START, Class: TestClassAFiveMethodsWithNoDepsSample, Class instance hash code: 249684623, Method name: testMethodA, Time of event: 2017-09-10 16:55:41.638, Thread ID: 5447}

Event logged at index 2: EventLog{Event: LISTENER_TEST_METHOD_START, Class: TestClassAFiveMethodsWithNoDepsSample, Class instance hash code: 473286673, Method name: testMethodB, Time of event: 2017-09-10 16:55:41.640, Thread ID: 5448}
Event logged at index 3: EventLog{Event: LISTENER_TEST_METHOD_START, Class: TestClassAFiveMethodsWithNoDepsSample, Class instance hash code: 249684623, Method name: testMethodB, Time of event: 2017-09-10 16:55:41.645, Thread ID: 5449}

Event logged at index 4: EventLog{Event: LISTENER_TEST_METHOD_START, Class: TestClassAFiveMethodsWithNoDepsSample, Class instance hash code: 473286673, Method name: testMethodC, Time of event: 2017-09-10 16:55:41.657, Thread ID: 5450}
Event logged at index 5: EventLog{Event: LISTENER_TEST_METHOD_START, Class: TestClassAFiveMethodsWithNoDepsSample, Class instance hash code: 249684623, Method name: testMethodC, Time of event: 2017-09-10 16:55:41.689, Thread ID: 5451}

Event logged at index 6: EventLog{Event: LISTENER_TEST_METHOD_START, Class: TestClassAFiveMethodsWithNoDepsSample, Class instance hash code: 473286673, Method name: testMethodD, Time of event: 2017-09-10 16:55:41.694, Thread ID: 5452}
Event logged at index 7: EventLog{Event: LISTENER_TEST_METHOD_START, Class: TestClassAFiveMethodsWithNoDepsSample, Class instance hash code: 249684623, Method name: testMethodD, Time of event: 2017-09-10 16:55:41.713, Thread ID: 5453}

Event logged at index 8: EventLog{Event: LISTENER_TEST_METHOD_START, Class: TestClassAFiveMethodsWithNoDepsSample, Class instance hash code: 249684623, Method name: testMethodE, Time of event: 2017-09-10 16:55:41.778, Thread ID: 5455}
Event logged at index 9: EventLog{Event: LISTENER_TEST_METHOD_START, Class: TestClassAFiveMethodsWithNoDepsSample, Class instance hash code: 473286673, Method name: testMethodE, Time of event: 2017-09-10 16:55:41.788, Thread ID: 5454}

**Event logs extracted from event log list between index 10 and index 19 should be the test method execution event logs for a block of 10 simultaneously executing methods**

Event logged at index 10: EventLog{Event: TEST_METHOD_EXECUTION, Class: TestClassAFiveMethodsWithNoDepsSample, Class instance hash code: 473286673, Method name: testMethodA, Time of event: 2017-09-10 16:55:42.117, Thread ID: 5446}
Event logged at index 11: EventLog{Event: TEST_METHOD_EXECUTION, Class: TestClassAFiveMethodsWithNoDepsSample, Class instance hash code: 249684623, Method name: testMethodA, Time of event: 2017-09-10 16:55:42.138, Thread ID: 5447}

Event logged at index 12: EventLog{Event: TEST_METHOD_EXECUTION, Class: TestClassAFiveMethodsWithNoDepsSample, Class instance hash code: 473286673, Method name: testMethodB, Time of event: 2017-09-10 16:55:42.140, Thread ID: 5448}
Event logged at index 13: EventLog{Event: TEST_METHOD_EXECUTION, Class: TestClassAFiveMethodsWithNoDepsSample, Class instance hash code: 249684623, Method name: testMethodB, Time of event: 2017-09-10 16:55:42.145, Thread ID: 5449}

Event logged at index 14: EventLog{Event: TEST_METHOD_EXECUTION, Class: TestClassAFiveMethodsWithNoDepsSample, Class instance hash code: 473286673, Method name: testMethodC, Time of event: 2017-09-10 16:55:42.157, Thread ID: 5450}
Event logged at index 15: EventLog{Event: TEST_METHOD_EXECUTION, Class: TestClassAFiveMethodsWithNoDepsSample, Class instance hash code: 249684623, Method name: testMethodC, Time of event: 2017-09-10 16:55:42.190, Thread ID: 5451}

Event logged at index 16: EventLog{Event: TEST_METHOD_EXECUTION, Class: TestClassAFiveMethodsWithNoDepsSample, Class instance hash code: 473286673, Method name: testMethodD, Time of event: 2017-09-10 16:55:42.194, Thread ID: 5452}
Event logged at index 17: EventLog{Event: TEST_METHOD_EXECUTION, Class: TestClassAFiveMethodsWithNoDepsSample, Class instance hash code: 249684623, Method name: testMethodD, Time of event: 2017-09-10 16:55:42.214, Thread ID: 5453}

Event logged at index 18: EventLog{Event: LISTENER_TEST_METHOD_PASS, Class: TestClassAFiveMethodsWithNoDepsSample, Class instance hash code: 473286673, Method name: testMethodA, Time of event: 2017-09-10 16:55:42.217, Thread ID: 5446}
Event logged at index 19: EventLog{Event: LISTENER_TEST_METHOD_PASS, Class: TestClassAFiveMethodsWithNoDepsSample, Class instance hash code: 249684623, Method name: testMethodA, Time of event: 2017-09-10 16:55:42.238, Thread ID: 5447}

**Event logs extracted from event log list between index 20 and index 29 should be the test method pass event logs for a block of 10 simultaneously executing methods**

Event logged at index 20: EventLog{Event: LISTENER_TEST_METHOD_PASS, Class: TestClassAFiveMethodsWithNoDepsSample, Class instance hash code: 473286673, Method name: testMethodB, Time of event: 2017-09-10 16:55:42.240, Thread ID: 5448}
Event logged at index 21: EventLog{Event: LISTENER_TEST_METHOD_PASS, Class: TestClassAFiveMethodsWithNoDepsSample, Class instance hash code: 249684623, Method name: testMethodB, Time of event: 2017-09-10 16:55:42.245, Thread ID: 5449}

Event logged at index 22: EventLog{Event: LISTENER_TEST_METHOD_PASS, Class: TestClassAFiveMethodsWithNoDepsSample, Class instance hash code: 473286673, Method name: testMethodC, Time of event: 2017-09-10 16:55:42.257, Thread ID: 5450}

Event logged at index 23: EventLog{Event: TEST_METHOD_EXECUTION, Class: TestClassAFiveMethodsWithNoDepsSample, Class instance hash code: 249684623, Method name: testMethodE, Time of event: 2017-09-10 16:55:42.278, Thread ID: 5455}
Event logged at index 24: EventLog{Event: TEST_METHOD_EXECUTION, Class: TestClassAFiveMethodsWithNoDepsSample, Class instance hash code: 473286673, Method name: testMethodE, Time of event: 2017-09-10 16:55:42.288, Thread ID: 5454}

Event logged at index 25: EventLog{Event: LISTENER_TEST_METHOD_PASS, Class: TestClassAFiveMethodsWithNoDepsSample, Class instance hash code: 249684623, Method name: testMethodC, Time of event: 2017-09-10 16:55:42.290, Thread ID: 5451}

Event logged at index 26: EventLog{Event: LISTENER_TEST_METHOD_PASS, Class: TestClassAFiveMethodsWithNoDepsSample, Class instance hash code: 473286673, Method name: testMethodD, Time of event: 2017-09-10 16:55:42.294, Thread ID: 5452}
Event logged at index 27: EventLog{Event: LISTENER_TEST_METHOD_PASS, Class: TestClassAFiveMethodsWithNoDepsSample, Class instance hash code: 249684623, Method name: testMethodD, Time of event: 2017-09-10 16:55:42.314, Thread ID: 5453}

Event logged at index 28: EventLog{Event: LISTENER_TEST_METHOD_PASS, Class: TestClassAFiveMethodsWithNoDepsSample, Class instance hash code: 249684623, Method name: testMethodE, Time of event: 2017-09-10 16:55:42.379, Thread ID: 5455}
Event logged at index 29: EventLog{Event: LISTENER_TEST_METHOD_PASS, Class: TestClassAFiveMethodsWithNoDepsSample, Class instance hash code: 473286673, Method name: testMethodE, Time of event: 2017-09-10 16:55:42.389, Thread ID: 5454}

juherr commented:
@selenareneephillips 

> So, my question is -- this has nothing to do with TestNG, right? All of this is due to Java's handling of the threads and it's not anything TestNG has control over? If so, then I can refactor my code pretty easily to process the logs differently.

Difficult to say. I think it is related to the power of the computer and available resources. Depending on the worker type, we have some random issues on TeamCity too.
Except by using some Docker stuff with limited resources, I don't know how to reproduce the issue.
So, if you can update the tests and still check the feature, go ahead! :)

juherr commented:
Ping @cbeust

cbeust commented:
@juherr You ok with merging this?

selenareneephillips commented:
Wait. I am working on a refactor. I'll close this PR and open a new one.
